File: enode/src/main/java/org/enodeframework/eventing/impl/DefaultEventSerializer.java
package org.enodeframework.eventing.impl;

import com.google.common.collect.Maps;
import org.enodeframework.common.serializing.ISerializeService;
import org.enodeframework.eventing.IDomainEvent;
import org.enodeframework.eventing.IEventSerializer;
import org.enodeframework.infrastructure.ITypeNameProvider;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * @author anruence@gmail.com
 */
public class DefaultEventSerializer implements IEventSerializer {

    private final ITypeNameProvider typeNameProvider;

    private final ISerializeService serializeService;

    public DefaultEventSerializer(ITypeNameProvider typeNameProvider, ISerializeService serializeService) {
        this.typeNameProvider = typeNameProvider;
        this.serializeService = serializeService;
    }

    @Override
    public Map<String, String> serialize(List<IDomainEvent<?>> evnts) {
        LinkedHashMap<String, String> dict = Maps.newLinkedHashMap();
        evnts.forEach(evnt -> {
            String typeName = typeNameProvider.getTypeName(evnt.getClass());
            String eventData = serializeService.serialize(evnt);
            dict.put(typeName, eventData);
        });
        return dict;
    }

    @Override
    public List<IDomainEvent<?>> deserialize(Map<String, String> data) {
        List<IDomainEvent<?>> evnts = new ArrayList<>();
        data.forEach((key, value) -> {
            Class<?> eventType = typeNameProvider.getType(key);
            IDomainEvent<?> evnt = (IDomainEvent<?>) serializeService.deserialize(value, eventType);
            evnts.add(evnt);
        });
        return evnts;
    }
}

File: enode/src/main/java/org/enodeframework/eventing/impl/DefaultEventSerializer.kt
package org.enodeframework.eventing.impl

import org.enodeframework.common.serializing.SerializeService
import org.enodeframework.eventing.DomainEventMessage
import org.enodeframework.eventing.EventSerializer
import org.enodeframework.infrastructure.TypeNameProvider

/**
 * @author anruence@gmail.com
 */
class DefaultEventSerializer(
    private val typeNameProvider: TypeNameProvider,
    private val serializeService: SerializeService
) : EventSerializer {
    override fun serialize(evnts: List<DomainEventMessage>): Map<String, String> {
        return evnts.associateBy({ k -> typeNameProvider.getTypeName(k.javaClass) },
            { v -> serializeService.serialize(v) })
    }

    override fun deserialize(data: Map<String, String>): List<DomainEventMessage> {
        return data.map { (key, value) ->
            val eventType = typeNameProvider.getType(key)
            serializeService.deserialize(value, eventType) as DomainEventMessage
        }
    }
}